### ISSUE TITLE
Pixie VM ver 2 and Pixie-in-Pixie

### ISSUE THREAD

**Issue by halgari** [pixie-lang/pixie#312]
This is going to be a massive undertaking, so let's get started. 

Problem:

1) The current VM uses stacklets heavily. This not only limits how we can run Pixie code (make run_interactive doesn't run with pypy), but also introduces porting problems (we can't port Pixie to platforms that don't support stacklets). 

2) Stacklets aren't exactly cheap to create, they aren't super expensive, but they do take about 1KB per stacklet.

3) Stacklets are black-boxes. How do they work? Well thats buried inside quite a bit of assembly and C. That I  don't really understand.  

4) The JIT can't dig into stacklets and optimize them much. Stacklet switching isn't super cheap either.  

Ratonale:

A somewhat recent paper published on Racket in RPython (`http://wphomes.soic.indiana.edu/jsiek/files/2014/04/pycket.pdf`), shows that with careful implementation of a VM, it's possible to implement `call-cc` inside RPython without the use of stacklets. Racket has an advantage over Pixie here since Racket's core libraries are quite a bit simpler than Pixie. Pixie requires vectors and hashmaps to even be able to read `stdlib.pxi`.

A previous attempt (see the pixie effects POC in this repo) showed that RPython can handle AST interpreters that make heavy use of trampolines, but we quickly overwhelmed the JIT on that approach do to trying to go too far and make every side effecting call an effect (and therefore a continuation). This approach could be simplified by keeping the Pixie semantics the same and simply move to a CPS for the interpreter and VM code, while not trying to make these continuations side-effect free. This approach worked well for Pycket and should work well for Pixie as well. 

Sadly, we cannot simply enable CPS in our existing code. We would have to hand re-write all the collection classes, reader, and compiler in CPS form. That would take a lot of work and result in code that is completely unmaintainable. Instead it's simpler to leverage compilers to do this work for us. 

Solution:
The goal of this effort will be to reduce the code found in pixie.vm.* to the smallest sub-set of Pixie.  We will then re-write the collections, compiler, and reader in Pixie. We can use pixie to compile these functions to RPython (using a special RPython backend for our compiler). 

As with all bootstrapped languages the problem will then be that we need pixie to compile pixie. Therefore we'll need to check in the RPython code into Git. Thus the generated RPython code should be semi git friendly. 

The RPython generated by the pixie compiler will simply be the AST constructors. Thus, the generated RPython code for the form `(def foo 1)` will look something like this:

```python

ast.fn(
  name="__toplevel__fn",
  args=[],
  body=ast.Do(
    ast.Def(
      name="user/Foo",
      val=ast.Const(
        rt.wrap(1)
      )
    )
  )
)
 
```

File Changes: 

Below is a file-by-file overview of the actions needed to move to the new vm format. Notes on each file are as follows: migrate = move the file with minimal changes to the new vm. port = rewrite the file in Pixie (and compile it to RPython). delete = not needed, remove file. rewrite = will remain as rpython but will need to be rewritten with new logic. 

```
array.py - migrate
atom.py - port
bits.py - migrate
bootstrap.py - delete
code.py - migrate
compiler.py - port
cons.py - port
custom_types.py - migrate
interpreter.py - rewrite
keyword.py - migrate
lazy_seq.py - port
map_entry.py - port
numbers.py - migrate
object.py - migrate/rewrite (exception support will need to be rewritten)
persistent_*.py - port
primitives.py - migrate
reader.py - port
reduced.py - port
rt.py - rewrite
stacklet.py - delete (will be replaced with new interpreter features)
stdlib.py - port
string.py - migrate
string_builder.py - migrate
symbol.py - migrate
threads.py - migrate
util.py - migrate

```

**Comment by andrewchambers** [comment]
Is this how chicken scheme does it's implementation?

**Comment by andrewchambers** [comment]
To be clear, this will allow proper tail recursion, green threads and change how FFI needs to be done right?

**Comment by heyLu** [comment]
There seems to be a [followup paper](http://homes.soic.indiana.edu/samth/pycket-draft.pdf) now, by some of the same authors.

**Comment by andrewchambers** [comment]
cool https://github.com/samth/pycket

**Comment by halgari** [comment]
Yes, this will enable proper tail recursion, we have call/cc but it's based on copying the C stack. In this new implementation the stack of a pixie program is just a linked-list of continuations. So if you need to, you can copy, splice, or do whatever you want to the stack. That means that creating a new lightweight thread is about as cheap as a single lazy-seq cell. Switching between lightweight threads is very, cheap, and it's possible to perhaps hint to the JIT to optimize thread switching (still need to think about this one). 

FFI won't change. 

We will no longer be able to create a call/cc continuation that calls through an ffi function, however. So yo won't be able to call qsort and expect to get a full continuation from inside the callback. But I think use cases for that would be pretty rare.

**Comment by halgari** [comment]
And yes, that paper is good, I've taken a few hints from it in implementing this interpreter.

**Comment by andrewchambers** [comment]
The super lightweight coroutines would make a runtime system similar to Go or even erlang quite interesting.

**Comment by brandonbloom** [comment]
I totally support this design approach, however, I suggest you use this opportunity to *eliminate undelimited call-cc*. @halgari, you and I have discussed this privately before, but for the benefit of other parties I'll pretend like we haven't.

Pixie's current implementation frequently uses call-cc to captures all stack frames, rather than a range of stack-frames to some delimiter. Because of this, all of the current async, yield, and similar in Pixie are very prone to memory leaks. It need not be that way. See http://okmij.org/ftp/continuations/against-callcc.html

Skimming your usages of call-cc, most of them can pretty easily be replaced with one of a small handful of alternatives built on delimited continuations. If you don't want to do a full effects system with handlers and the like, you should choose to either 1) built upon generalized yield or 2) implement [multi-prompt delimited continuation operators][2] directly.

For (1), see www.cs.indiana.edu/~sabry/papers/yield.pdf
I can't dig up my favorite reference for (2) at the moment, but will let you know if I find it.

**Comment by halgari** [comment]
I think there's a bit of misunderstanding of how `call-cc` works in Pixie. And perhaps that's simply because I call it `call-cc` when it really isn't. 

As the `against-cc` article states, there is no such thing as full continuations. And this is true in Pixie. In Pixie, we have single-shot continuations that are delimited by the stacklet they are called in. The VM kicks off inside a default stacklet, from there users can create new "threads" by calling `-new-stacklet`. This creates a new stack and executes a function in that stack. `call-cc` then creates a single-use function starting from the current point of execution to the end of the function handed to `-new-stacklet`. To me that sounds a lot like delimited continuations. 

@brandonbloom or perhaps I'm missing something here.

**Comment by brandonbloom** [comment]
Ah, OK, I understand our respective misunderstandings now :-)

Your call-cc really is call-cc, but it's delimited by the stacklets, which are akin to lightweight threads. This is essentially what oleg means by saying undelimited continuations don't really exist in practice. There's always a delimiter of some kind, minimally the process boundary! I misunderstood the code in pixie because I saw the full stack copy and not the new-stack creation code: You don't have the *bad* memory leak problem, but you're definitely missing the expressiveness of delimited continuations.

What Pixie is missing is *multi-prompt* delimited continuations. Take a look at the types and operators defined in ["Delimited Control in OCaml, Abstractly and Concretely"][1] on the bottom of page 4. If you go with the exception handling analogy, `new_prompt` is similar to dynamically creating a class `MyExceptionType`, and `push_prompt` is similar to `catch MyExceptionType`. Then `throw` is just `take_subcont` and discarding the value. If you didn't discard the value, you could have a restartable exception via `push_subcont`.

Going with the yield analogy, what you have is `(defn yield [x] ...)` to generate a value, but that can be defined in terms of `(defn yield-to [who x] ...)` as `(defn yield [x] (yield-to *whoever* x))`.

Hopefully between Oleg's paper and this explaination, it will be clear :-) As always, keep up the awesome work!

[1]: http://okmij.org/ftp/continuations/implementations.html#delimcc-paper

**Comment by andrewchambers** [comment]
@halgari Looking quickly through the vm2 branch, has the plan changed? It doesn't look like you are emitting rpython anymore.  Could you explain which parts are written in pixie and which parts must remain in rpython? I am a bit confused about what is being bootstrapped by what.

**Comment by halgari** [comment]
Update on everything, it's been some time, but I'm making progress. Whatever happens I have to get this done before the end of Sep since I'm giving a talk Strangeloop on it at that time. 

So progress update:

We have compiler.pxi. With the old pixie vm we can compile that compiler, complete with a bootstrap env, and feed it into VM2. The compilation process is found in compile-bootstrap.pxi and it works as follows:

Load the compiler, pxic-writer.pxi and compile bootstrap.pxi, bootstrap-macros.pxi, compiler.pxi, ast.pxi, etc. This results in bootstrap.pxic, a 600KB file that contains all the data structure code, all of pixie.stdlib, compiler, etc. Now you can run VM2, that will load bootstrap.pxic and dump you at a repl. This resulting repl uses the compiled compiler, and that's where stuff gets fun. 

Currently I'm trying to load ffi-infer.pxi on the compiled compiler (completely bootstrapped VM2), it's moving along, but a lot of bug fixes are needed. After I get ffi-infer, I can get the libuv stuff loaded, then pixie.fs, and then pixie.test. 

So all I can say is keep up the good work! Once I get the VM2 stable, I'll start running tests. Those failing tests will expose where I need to add back in functionality I missed (e.g. BigInts aren't in VM2 yet, so all the recent work around that will cause failing tests). Those missing bits will either need to be ported to the new VM or rewritten in pixie. 

As it stands, the new VM is very light. It compiles without a JIT in about 120sec on my box. And because it is so small and light we can easily add improvements without having to rewrite whole parts of the VM.

**Comment by xkxx** [comment]
@halgari any updates on vm2? Been looking forward to it for a while.

**Comment by halgari** [comment]
I worked a ton on this, but ran into time issues before the Strangloop conference. At Strangeloop I got talking to one of the authors of Pycket. Unfortunately it sounds like their implementation isn't much better than mine. The amount of load that is put on the JIT due to this sort of interpreter is quite large. In short, I'd rather have a mutable fast interpreter than a immutable `mostly fast` interpreter. Shelving this work for now. The PR will still exist, but I'm not convinced the features are worth the cost.
